### PR TITLE
VPC Flow Logs enforcer

### PR DESCRIPTION
**Pull request by apsureda**
Implementation of a Cloud Function that will automatically enforce VPC Flow logs whenever a subnet is changed or modified.

### CODE REVIEW

File: tools/vpc-flowlogs-enforcer/terraform/templates/cloud_function/main.py [hunk 1]
```diff
@@ -0,0 +1,174 @@
+"""
+ Copyright 2020 Google LLC. This software is provided as-is, without warranty
+ or representation for any use or purpose. Your use of it is subject to your
+ agreement with Google.
+"""
+import logging
```
Comment (ludoo) [inline tools/vpc-flowlogs-enforcer/terraform/templates/cloud_function/main.py:6]: can you order import lexically?

File: tools/vpc-flowlogs-enforcer/terraform/templates/cloud_function/main.py [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+from dateutil import parser
+from googleapiclient import discovery
+
+
+def enable_net_logs(event, context):
+    """
+  Enable flow logs in a subnetwork using the log configuration provided in the
+  LOG_CONFIG environment variable.
+  """
+    default_log_config = '{"aggregationInterval" : "INTERVAL_5_SEC", ' \
```
Comment (ludoo) [inline tools/vpc-flowlogs-enforcer/terraform/templates/cloud_function/main.py:24]: I would put these as module-level constants

File: tools/vpc-flowlogs-enforcer/terraform/templates/cloud_function/main.py [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+        return
+
+    logs_enabled = False
+    project = region = subnetwork = None
+
+    # Initialize the compute API
+    service = discovery.build('compute', 'v1', cache_discovery=False)
+
+    # Stackdriver change notification
+    if 'resource' in json_msg and 'type' in json_msg['resource'] and json_msg[
```
Comment (ludoo) [inline tools/vpc-flowlogs-enforcer/terraform/templates/cloud_function/main.py:47]: I would prefer having this code split out into separate smaller functions, which take a json as input and can then be easily tested

File: tools/vpc-flowlogs-enforcer/terraform/templates/cloud_function/main.py [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+                              json_msg['asset']['name'])
+                if m:
+                    if elem == 'project':
+                        project = m.group(1)
+                    elif elem == 'region':
+                        region = m.group(1)
+                    elif elem == 'subnetwork':
+                        subnetwork = m.group(1)
+    else:
+        logging.error('evet type is not gce_subnetwork. Ignoring event: %s',
```
Comment (ludoo) [inline tools/vpc-flowlogs-enforcer/terraform/templates/cloud_function/main.py:77]: typo on `evet`

File: tools/vpc-flowlogs-enforcer/terraform/templates/cloud_function/main.py [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+        response = get_request.execute()
+    except Exception as e:  # pylint: disable=broad-except
+        # Replace new lines with spaces so as to prevent several entries which
+        # would trigger several errors.
+        error_message = traceback.format_exc().replace('\n', '  ')
+        logging.error('can\'t get subnet information: %s', error_message)
+        time.sleep(3)  # currently required because of b/155636171
+        raise e
+
+    # check if flow logs are enabled
```
Comment (ludoo) [inline tools/vpc-flowlogs-enforcer/terraform/templates/cloud_function/main.py:102]: I would also split this code into separate testable functions

File: tools/vpc-flowlogs-enforcer/terraform/templates/cloud_function/main.py [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+import re
+import os
+import sys
+import time
+from datetime import datetime, timezone
+from dateutil import parser
+from googleapiclient import discovery
+
+
+def enable_net_logs(event, context):
```
Comment (ludoo) [inline tools/vpc-flowlogs-enforcer/terraform/templates/cloud_function/main.py:19]: If you restructure the module into smaller functions, this function then will just hold the core logic and should imho be identified as the main entry point, by renaming it to `main`

**Review by ludoo — COMMENTED** [review]
First pass, just on the module's structure

**Comment by apsureda** [comment]
I made all the proposed changes. Can you take a look again?

File: tools/vpc-flowlogs-enforcer/README.md [hunk 7]
```diff
@@ -62,7 +62,7 @@ pip3 install -r terraform/templates/cloud_function/requirements.txt
 Run the cloud funtion using any of the sample logs provided (or create your own):
```
Comment (ludoo) [inline tools/vpc-flowlogs-enforcer/README.md:62]: typo :)

**Review by ludoo — APPROVED** [review]
Thanks Alfonso :)

**Pull request merged**